src/storage-controller/src/lib.rs

```diff
@@ -2657,21 +2657,41 @@ where
                         .await
                         .expect("invalid persist usage");
 
-                    if !write_handle.upper().is_empty() {
-                        write_handle
-                            .append(
-                                Vec::<((mz_storage_types::sources::SourceData, ()), T, Diff)>::new(
-                                ),
-                                write_handle.upper().clone(),
-                                Antichain::new(),
-                            )
-                            .await
+                    if write_handle.upper().is_empty() {
+                        Some(shard_id)
+                    } else {
+                        // Finalizing a shard can take a long time cleaning up existing data.
+                        // Spawning a task means that we can't proactively remove this shard
+                        // from the finalization register, unfortunately... but the next run
+                        // of `finalize_shards` should notice the upper has advanced and tidy
+                        // up.
+                        mz_ore::task::spawn(|| format!("finalize_shard({shard_id})"), async move {
+                            let result =
+                                write_handle
+                                    .append(
+                                        Vec::<(
+                                            (mz_storage_types::sources::SourceData, ()),
+                                            T,
+                                            Diff,
+                                        )>::new(),
+                                        write_handle.upper().clone(),
+                                        Antichain::new(),
+                                    )
+                                    .await;
+
                             // Rather than error, just leave this shard as one to finalize later.
-                            .ok()?
-                            .ok()?;
+                            match result {
+                                Err(usage) => {
+                                    error!("invalid usage while finalizing shard: {usage:?}")
+                                }
+                                Ok(Err(mismatch)) => {
```

Comment:
This call is using `append`, which automatically retries if the upper advances. Looks like `append` returns this error if you called it with an upper that is too far ahead of the actual one (e.g. the real upper is 5, but your append call is `(7, ())`. We're using `write_handle.upper().clone()` above, which means this should never happen and we could probably be more aggressive here if we liked (i.e. error or panic)

Reply:
Point! I bumped to `warn` for now, but I'd be happy to make it even more aggressive in the future.